compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalMatmul.cpp

```diff
@@ -0,0 +1,211 @@
+    producer = producerOutput.getDefiningOp();
+    if (!producer) {
+      return failure();
+    }
+    if (!matchPattern(producer, m_Op<arith::NegFOp>())) {
+      return failure();
+    }
+  }
+
+  return success();
+}
+
+static LogicalResult isGenericMultiplyOp(linalg::GenericOp genericOp) {
+  // Check that all loops are parallel
+  unsigned numLoops = genericOp.getNumLoops();
+  unsigned numParallelLoops = genericOp.getNumParallelLoops();
+  if (numLoops != numParallelLoops) {
+    return failure();
+  }
+
+  // Work back from linalg.yield and check body of genericOp.
+  // The genericOp should yield the result of an arith.mulf.
+  auto yieldOp = cast<linalg::YieldOp>(genericOp.getBody()->getTerminator());
+  Value producerOutput;
+  Operation *producer;
+
+  // Producer of linalg.yield op is input
+  {
+    producerOutput = yieldOp->getOperand(0);
+    producer = producerOutput.getDefiningOp();
+    if (!producer || producer->getNumOperands() == 0) {
+      return failure();
+    }
+    if (!matchPattern(producer, m_Op<arith::MulFOp>())) {
+      return failure();
+    }
+  }
+
+  return success();
+}
+
+// This pattern does a basic fusion of matmul, silu, matmul
+// ops, moving them into a single `flow.dispatch.region` op.
+class FuseHorizontalMatmulPattern final
+    : public OpRewritePattern<linalg::GenericOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    // Fail if not the right type of linalg.generic
+    if (isGenericMultiplyOp(genericOp).failed())
+      return failure();
+
+
+    auto genericMultiplyOp = genericOp->getOperand(0).getDefiningOp<linalg::GenericOp>();
+    auto matmulOp1 = genericOp->getOperand(1).getDefiningOp<linalg::MatmulTransposeBOp>();
+    if (!genericMultiplyOp || !matmulOp1 || isGenericMultiplyOp(genericMultiplyOp).failed())
+      return failure();
+    
+    auto genericSigmoidOp = genericMultiplyOp->getOperand(0).getDefiningOp<linalg::GenericOp>();
+    auto matmulOp2 = genericMultiplyOp->getOperand(1).getDefiningOp<linalg::MatmulTransposeBOp>();
+    if (!genericSigmoidOp || !matmulOp2)
+      return failure();
+    if (!genericSigmoidOp || !matmulOp2 || isGenericSigmoidOp(genericSigmoidOp).failed())
+      return failure();
+    auto genericSigmoidOpInput = genericSigmoidOp->getOperand(0).getDefiningOp<linalg::MatmulTransposeBOp>();
+    if (!genericSigmoidOpInput || genericSigmoidOpInput != matmulOp2)
+      return failure();
+
+    auto matmulOp1Input = matmulOp1->getOperand(0).getDefiningOp<tensor::CollapseShapeOp>();
+    auto matmulOp2Input = matmulOp2->getOperand(0).getDefiningOp<tensor::CollapseShapeOp>();
+    if (!matmulOp1Input || !matmulOp2Input || matmulOp1Input != matmulOp2Input)
+      return failure();
+
+    // Fail if matmul is already in a dispatch
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(genericOp)) {
```

Comment:
Might be good to check that all the other ops are not in a dispatch here as well